doc/source/rllib/rllib-callback.rst

```diff
@@ -0,0 +1,363 @@
+over individual or lists of callables provided through the various arguments of
+the :py:meth:`~ray.rllib.algorithms.algorithm_config.AlgorithmConfig.callbacks` method.
+
+For example, assume the callback event is ``on_train_result``, which fires at the end of
+a training iteration and inside the algorithm's process.
+
+- RLlib loops through the list of all given :py:class:`~ray.rllib.callbacks.callbacks.RLlibCallback`
+  subclasses and calls their ``on_train_result`` method. Thereby, it keeps the exact order the user
+  provided in the list.
+- RLlib then loops through the list of all defined ``on_train_result`` callables. The user configured these
+  by calling the :py:meth:`~ray.rllib.algorithms.algorithm_config.AlgorithmConfig.callbacks` method
+  and defining the ``on_train_result`` argument in this call.
+
+.. code-block:: python
+
+    class MyCallbacks(RLlibCallback):
+        def on_train_result(self, *, algorithm, metrics_logger, result, **kwargs):
+            print("RLlibCallback subclass")
+
+    class MyDebugCallbacks(RLlibCallback):
+        def on_train_result(self, *, algorithm, metrics_logger, result, **kwargs):
+            print("debug subclass")
+
+    # Define the callbacks order through the config.
+    # Subclasses first, then individual `on_train_result` (or other events) callables:
+    config.callbacks(
+        callbacks_class=[MyDebugCallbacks, MyCallbacks],  # <- note: debug class first
+        on_train_result=[
+            lambda algorithm, **kw: print('lambda 1'),
+            lambda algorithm, **kw: print('lambda 2'),
+        ],
+    )
+
+    # When training the algorithm, after each training iteration, you should see
+    # something like:
+    # > debug subclass
+    # > RLlibCallback subclass
+    # > lambda 1
+    # > lambda 2
+
+
+Examples
+--------
+
+Here are two examples showing you how to setup custom callbacks on the
+:ref:`Algorithm <rllib-key-concepts-algorithms>` process as well as on the
+:ref:`EnvRunner <rllib-key-concepts-env-runners>` processes.
+
+
+Example 1: `on_train_result`
+----------------------------
+
+The following example demonstrates how to implement a simple custom function writing the replay buffer
+contents to disk from time to time.
+
+You normally don't want to write the contents of buffers along with your
+:ref:`Algorithm checkpoints <rllib-checkpointing-docs>`, so doing this less often, in a more
+controlled fashion through a custom callback could be a good compromise.
+
+.. testcode::
+
+    import ormsgpack
+    from ray.rllib.algorithms.dqn import DQNConfig
+
+    def _write_buffer_if_necessary(algorithm, metrics_logger, result):
+        # Write the buffer contents only every ith iteration.
+        if algorithm.training_iteration % 2 == 0:
+            # python dict
+            buffer_contents = algorithm.local_replay_buffer.get_state()
+
+            # binary
+            msgpacked = ormsgpack.packb(
```

Comment:
Why do we use `ormsgpack` instead of `msgpack`?

Reply:
Good question. This is still in an experimental phase.
`ormsgpack` is apparently faster and supports numpy natively (we don't need an extra install, like `msgpack-numpy`).
It might(!) create larger checkpoint files, but we'll see.